spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/msi/ConfigMSICredentials.java

```diff
@@ -0,0 +1,196 @@
+        this.resource = environment.managementEndpoint();
+        this.msiType = checkMSIType();
+        this.metadataService = getMetadataService(msiType);
+        setHeader(msiType);
+    }
+
+    @Override
+    public String getToken(String resource) {
+        return this.getTokenFromIMDSEndpoint(resource == null ? this.resource : resource);
+    }
+
+    private String getTokenFromIMDSEndpoint(String resource) {
+        MSIToken token = cache.get(resource);
+        if (token != null && !token.isExpired()) {
+            return token.accessToken();
+        }
+        lock.lock();
+        try {
+            token = cache.get(resource);
+            if (token != null && !token.isExpired()) {
+                return token.accessToken();
+            }
+
+            token = retrieveTokenFromIDMSWithRetry(resource);
+            if (token != null) {
+                cache.put(resource, token);
+            }
+            return token.accessToken();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private MSIToken retrieveTokenFromIDMSWithRetry(String resource) {
+        HttpGet httpGet = new HttpGet(this.metadataService.buildTokenUrl(resource));
+        this.tokenReqHeaders.forEach(httpGet::setHeader);
+
+        CloseableHttpResponse response = null;
+
+        int retry = 0;
+        while (retry < MAX_RETRY) {
+            try {
+                response = HTTP_CLIENT.execute(httpGet);
+                return ADAPTER.deserialize(toString(response), MSIToken.class);
+            } catch (Exception e) {
+                LOGGER.error(RETRIEVE_IMDS_FAILURE_MSG, e);
+
+                if (canRetry(response)) {
+                    retryResponse(response, retry++);
+                } else {
+                    throw new IllegalStateException(RETRIEVE_IMDS_FAILURE_MSG);
+                }
+            } finally {
+                if (response != null) {
+                    try {
+                        response.close();
+                    } catch (IOException e) {
+                        LOGGER.warn("Failed to close the response", e);
+                    }
+                }
+            }
+        }
+
+        throw new IllegalStateException(RETRIEVE_IMDS_FAILURE_MSG);
+    }
+
+    private String toString(CloseableHttpResponse response) {
+        StringWriter writer = new StringWriter();
+        try {
+            IOUtils.copy(response.getEntity().getContent(), writer, "UTF-8");
+        } catch (IOException e) {
+            LOGGER.warn("Failed to convert response {} to String.", response, e);
+        }
+        return writer.toString();
+    }
+
+    private boolean canRetry(CloseableHttpResponse response) {
+        if (response == null) {
+            return false;
+        }
+
+        int statusCode = response.getStatusLine().getStatusCode();
+        return statusCode == 404 || statusCode == 429 || (statusCode >= 500 && statusCode <= 599);
```

Comment:
Let's use `HttpStatus` instead of the actual code number.

Reply:
Will keep it as-is for now.